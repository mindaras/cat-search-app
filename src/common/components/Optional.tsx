interface Props {
  if: boolean;
  children: any;
}

const Optional: React.FC<Props> = (props) => {
  if (props.if) {
    return props.children;
  }

  return null;
};

export { Optional };
