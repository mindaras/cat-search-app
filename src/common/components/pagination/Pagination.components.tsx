import styled from "styled-components";

interface ContainerProps {
  disabled: boolean;
}

const Container = styled.div<ContainerProps>`
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 0 20px;
  height: ${(p) => p.theme.sizes.pagination.height};
  opacity: ${(p) => (p.disabled ? "0.3" : "1")};
`;

const Content = styled.div`
  display: flex;
  align-items: flex-end;
`;

interface ButtonProps {
  active: boolean;
}

const Button = styled.div<ButtonProps>`
  background-color: ${(p) =>
    p.active ? p.theme.colors.gray.medium : p.theme.colors.gray.light};
  border-radius: 100%;
  border: none;
  width: 40px;
  height: 40px;
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  padding: 0;
  font-size: 12px;
  margin: 0 4px;
  transition: ease 0.3s;

  :hover {
    background-color: ${(p) => p.theme.colors.gray.medium};
  }
`;

const Dots = styled.div`
  margin: 0 4px;
  color: ${(p) => p.theme.colors.gray.dark};
`;

export { Container, Content, Button, Dots };
