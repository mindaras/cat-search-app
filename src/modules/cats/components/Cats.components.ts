import styled from "styled-components";

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: calc(
    100vh - ${(p) => p.theme.sizes.search.height} -
      ${(p) => p.theme.sizes.pagination.height}
  );
`;

export { LoaderContainer };
