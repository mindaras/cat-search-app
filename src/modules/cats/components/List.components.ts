import styled from "styled-components";

const Cat = styled.img`
  width: 25%;
  height: calc(100% / 3);
  object-fit: cover;
  border: 4px solid #fff;
  border-radius: 30px;

  ${(p) => p.theme.media.medium} {
    width: 33.33%;
    height: 30vw;
  }

  ${(p) => p.theme.media.small} {
    width: 100%;
    height: 100vw;
  }
`;

interface ContainerProps {
  single: boolean;
}

const Container = styled.div<ContainerProps>`
  position: relative;
  display: flex;
  flex-wrap: wrap;
  width: 100%;
  height: calc(
    100vh - ${(p) => p.theme.sizes.search.height} -
      ${(p) => p.theme.sizes.pagination.height}
  );

  ${(p) => {
    if (p.single) {
      return `
            justify-content: center;
            align-items: center;

            ${Cat} {
                width: 40vw;
                height: 40vw;

                ${p.theme.media.medium} {
                    width: 50vw;
                    height: 50vw;
                }

                ${p.theme.media.small} {
                    width: 80vw;
                    height: 80vw;
                }
            }
        `;
    }
  }};

  ${(p) => p.theme.media.medium} {
    height: auto;
  }
`;

const NoDataText = styled.div`
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  font-size: 24px;
`;

const NoDataContainer = styled.div`
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: calc(100vh - ${(p) => p.theme.sizes.search.height});

  ${Cat} {
    opacity: 0.1;
  }

  ${(p) => p.theme.media.small} {
    height: auto;
  }
`;

export { Cat, Container, NoDataContainer, NoDataText };
