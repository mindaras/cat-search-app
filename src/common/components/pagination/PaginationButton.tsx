import { Button } from "./Pagination.components";

interface Props {
  active: boolean;
  onClick: () => void;
  children: number;
}

const PaginationButton: React.FC<Props> = ({ active, onClick, children }) => (
  <Button active={active} onClick={onClick}>
    {children}
  </Button>
);

export { PaginationButton };
