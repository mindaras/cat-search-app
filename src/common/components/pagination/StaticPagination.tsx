import { Fragment } from "react";
import { PaginationButton } from "./PaginationButton";

interface Props {
  page: number;
  total: number;
  onChange: (page: number) => void;
}

const StaticPagination: React.FC<Props> = ({
  page: currentPage,
  total,
  onChange,
}) => (
  <Fragment>
    {new Array(total).fill("").map((_, page) => (
      <PaginationButton
        key={page}
        active={page === currentPage}
        onClick={() => onChange(page)}
      >
        {page + 1}
      </PaginationButton>
    ))}
  </Fragment>
);

export { StaticPagination };
