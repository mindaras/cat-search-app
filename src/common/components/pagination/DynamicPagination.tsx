import { Fragment } from "react";
import { Optional } from "../Optional";
import { Button, Dots } from "./Pagination.components";
import { PaginationButton } from "./PaginationButton";

interface Props {
  page: number;
  total: number;
  onChange: (page: number) => void;
}

const DynamicPagination: React.FC<Props> = ({
  page: currentPage,
  total,
  onChange,
}) => {
  const renderPages = () => {
    const isAtTheStart = currentPage < 3;
    const isAtTheEnd = currentPage > total - 4;
    const startingPage = !isAtTheStart ? currentPage - 1 : 1;
    const pages = isAtTheEnd
      ? [total - 4, total - 3, total - 2]
      : [startingPage, startingPage + 1, startingPage + 2];

    return (
      <Fragment>
        {pages.map((page) => (
          <PaginationButton
            key={page}
            active={page === currentPage}
            onClick={() => onChange(page)}
          >
            {page + 1}
          </PaginationButton>
        ))}
      </Fragment>
    );
  };

  return (
    <Fragment>
      <Button active={currentPage === 0} onClick={() => onChange(0)}>
        1
      </Button>
      <Optional if={currentPage > 2 && total > 7}>
        <Dots>...</Dots>
      </Optional>
      {renderPages()}
      <Optional if={currentPage < total - 3 && total > 7}>
        <Dots>...</Dots>
      </Optional>
      <Button
        active={currentPage === total - 1}
        onClick={() => onChange(total - 1)}
      >
        {total}
      </Button>
    </Fragment>
  );
};

export { DynamicPagination };
