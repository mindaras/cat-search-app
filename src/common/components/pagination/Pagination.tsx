import { Optional } from "../Optional";
import { DynamicPagination } from "./DynamicPagination";
import { Container, Content } from "./Pagination.components";
import { StaticPagination } from "./StaticPagination";

interface Props {
  page: number;
  total: number;
  disabled: boolean;
  onChange: (page: number) => void;
}

const Pagination: React.FC<Props> = ({
  page: currentPage,
  total,
  disabled,
  onChange,
}) => {
  const changeHandler = (page: number) => {
    if (disabled) return;
    onChange(page);
  };

  return (
    <Optional if={total > 1}>
      <Container disabled={disabled}>
        <Content>
          <Optional if={total > 7}>
            <DynamicPagination
              page={currentPage}
              total={total}
              onChange={changeHandler}
            />
          </Optional>
          <Optional if={total <= 7}>
            <StaticPagination
              page={currentPage}
              total={total}
              onChange={changeHandler}
            />
          </Optional>
        </Content>
      </Container>
    </Optional>
  );
};

export { Pagination };
