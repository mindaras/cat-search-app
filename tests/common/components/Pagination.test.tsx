import renderer from "react-test-renderer";
import { Container } from "@testUtils/components/Container";
import { Pagination } from "@common/components/pagination/Pagination";

describe("List", () => {
  it("shows 7 pages", () => {
    const component = renderer.create(
      <Container>
        <Pagination page={0} total={7} disabled={false} onChange={() => null} />
      </Container>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("shows 4 initial pages and the last page", () => {
    const component = renderer.create(
      <Container>
        <Pagination
          page={0}
          total={10}
          disabled={false}
          onChange={() => null}
        />
      </Container>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("shows first page, three middle pages and the last page", () => {
    const component = renderer.create(
      <Container>
        <Pagination
          page={4}
          total={10}
          disabled={false}
          onChange={() => null}
        />
      </Container>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("shows disabled state", () => {
    const component = renderer.create(
      <Container>
        <Pagination page={4} total={10} disabled={true} onChange={() => null} />
      </Container>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("doesn't show anything", () => {
    const component = renderer.create(
      <Container>
        <Pagination page={0} total={1} disabled={false} onChange={() => null} />
      </Container>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
