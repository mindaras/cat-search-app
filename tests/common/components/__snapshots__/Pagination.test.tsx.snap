// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List doesn't show anything 1`] = `null`;

exports[`List shows 4 initial pages and the last page 1`] = `
<div
  className="sc-bcXHqe VdVlc"
  disabled={false}
>
  <div
    className="sc-gswNZR jxmpEH"
  >
    <div
      className="sc-dkrFOg eVAtTd"
      onClick={[Function]}
    >
      1
    </div>
    <div
      className="sc-dkrFOg EGNnJ"
      onClick={[Function]}
    >
      2
    </div>
    <div
      className="sc-dkrFOg EGNnJ"
      onClick={[Function]}
    >
      3
    </div>
    <div
      className="sc-dkrFOg EGNnJ"
      onClick={[Function]}
    >
      4
    </div>
    <div
      className="sc-hLBbgP dhXWtN"
    >
      ...
    </div>
    <div
      className="sc-dkrFOg EGNnJ"
      onClick={[Function]}
    >
      10
    </div>
  </div>
</div>
`;

exports[`List shows 7 pages 1`] = `
<div
  className="sc-bcXHqe VdVlc"
  disabled={false}
>
  <div
    className="sc-gswNZR jxmpEH"
  >
    <div
      className="sc-dkrFOg eVAtTd"
      onClick={[Function]}
    >
      1
    </div>
    <div
      className="sc-dkrFOg EGNnJ"
      onClick={[Function]}
    >
      2
    </div>
    <div
      className="sc-dkrFOg EGNnJ"
      onClick={[Function]}
    >
      3
    </div>
    <div
      className="sc-dkrFOg EGNnJ"
      onClick={[Function]}
    >
      4
    </div>
    <div
      className="sc-dkrFOg EGNnJ"
      onClick={[Function]}
    >
      5
    </div>
    <div
      className="sc-dkrFOg EGNnJ"
      onClick={[Function]}
    >
      6
    </div>
    <div
      className="sc-dkrFOg EGNnJ"
      onClick={[Function]}
    >
      7
    </div>
  </div>
</div>
`;

exports[`List shows disabled state 1`] = `
<div
  className="sc-bcXHqe cPnbMQ"
  disabled={true}
>
  <div
    className="sc-gswNZR jxmpEH"
  >
    <div
      className="sc-dkrFOg EGNnJ"
      onClick={[Function]}
    >
      1
    </div>
    <div
      className="sc-hLBbgP dhXWtN"
    >
      ...
    </div>
    <div
      className="sc-dkrFOg EGNnJ"
      onClick={[Function]}
    >
      4
    </div>
    <div
      className="sc-dkrFOg eVAtTd"
      onClick={[Function]}
    >
      5
    </div>
    <div
      className="sc-dkrFOg EGNnJ"
      onClick={[Function]}
    >
      6
    </div>
    <div
      className="sc-hLBbgP dhXWtN"
    >
      ...
    </div>
    <div
      className="sc-dkrFOg EGNnJ"
      onClick={[Function]}
    >
      10
    </div>
  </div>
</div>
`;

exports[`List shows first page, three middle pages and the last page 1`] = `
<div
  className="sc-bcXHqe VdVlc"
  disabled={false}
>
  <div
    className="sc-gswNZR jxmpEH"
  >
    <div
      className="sc-dkrFOg EGNnJ"
      onClick={[Function]}
    >
      1
    </div>
    <div
      className="sc-hLBbgP dhXWtN"
    >
      ...
    </div>
    <div
      className="sc-dkrFOg EGNnJ"
      onClick={[Function]}
    >
      4
    </div>
    <div
      className="sc-dkrFOg eVAtTd"
      onClick={[Function]}
    >
      5
    </div>
    <div
      className="sc-dkrFOg EGNnJ"
      onClick={[Function]}
    >
      6
    </div>
    <div
      className="sc-hLBbgP dhXWtN"
    >
      ...
    </div>
    <div
      className="sc-dkrFOg EGNnJ"
      onClick={[Function]}
    >
      10
    </div>
  </div>
</div>
`;
